proposals/0000-existentials.rst

```diff
@@ -207,8 +217,17 @@ Proposed Change Specification
       We can implement this rule by using a fresh unification variable
       for ``witness_ty`` before checking ``e`` against ``inner_ty``.
 
-   #. ``Witness`` would need to be added to Core, as well, likely as a new
-      constructor of GHC's ``Type`` type.
+   #. Types in the type-checker are represented using the same GHC type ``Type``
+      as Core types. However, a witness type in Haskell will use a *Haskell*
+      expression, not a *Core* expression. It thus requires a different constructor
+      of ``Type``, which will not be used in Core programs (that is, it will always
+      be a core-lint error)::
+
+        data Type = ...
+                  | HsWitnessTy Type (HsExpr GhcTc)
```

Comment:
I don't like the implications of this on the module graph of GHC. Very implementation-specific comment of course. Maybe this will be the time to have a dedicated source Haskell Type type, and allow the two type languages to evolve separately?

Reply:
Quite right. Added as an Effect.